File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rows = size(grid);
        int cols = size(grid[0]); 

        int count = 0; 

        for (int r = 0; r < rows; r++) { 
            for (int c = 0; c < cols; c++) { 
                if (grid[r][c] == '1') { 
                    count++; 
                    dfs(r, c, grid); 
                }
            }
        }

        return count;
    }

    void dfs(int r, int c, vector<vector<char>>& grid) { 
        int rows = size(grid); 
        int cols = size(grid[0]); 
        if (r >= 0 && r < rows && c >= 0 && c < cols && grid[r][c] == '1' && grid[r][c] == '1') { 
                grid[r][c] = '0'; 
                dfs(r + 1, c, grid); 
                dfs(r - 1, c, grid); 
                dfs(r, c + 1, grid); 
                dfs(r, c - 1, grid);
            } 
    
    }
};

File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        res = 0 

        def dfs(r, c): 
            if not (0 <= r < rows and 0 <= c < cols and grid[r][c] == 1): 
                return 0 
            
            grid[r][c] = 0

            return (
                1 + 
                dfs(r + 1, c) + 
                dfs(r - 1, c) + 
                dfs(r, c + 1) + 
                dfs(r, c - 1)
            )


        for r in range(rows): 
            for c in range(cols): 
                if grid[r][c] == 1: 
                    res = max(res, dfs(r, c))

        return res 


        
            


        
        